python-checks/src/main/java/org/sonar/python/checks/SpecialMethodReturnTypeCheck.java

```diff
@@ -0,0 +1,361 @@
+    }
+  }
+
+  private static void isGetNewArgsExCompliant(SubscriptionContext ctx, ReturnStatement returnStatement) {
+    List<Expression> returnedExpressions = returnStatement.expressions();
+    int numReturnedExpressions = returnedExpressions.size();
+
+    // If there is only one expression being returned, it might be a tuple wrapped in parentheses.
+    // I.e.
+    //
+    // return a, b
+    //
+    // and
+    //
+    // return (a, b)
+    //
+    // both return a tuple of two elements.
+    //
+    // We check for the second case and unwrap it:
+    if (numReturnedExpressions == 1) {
+      final Expression firstExpression = returnedExpressions.get(0);
+      if (firstExpression instanceof TupleImpl) {
+        // If a single expression is being returned, and it is a tuple, we directly inspect its elements:
+        returnedExpressions = ((TupleImpl) firstExpression).elements();
+        numReturnedExpressions = returnedExpressions.size();
+      } else {
+        // If there is only one expression being returned, and it is not a tuple expression, then
+        // we can not tell if it is a compliant tuple without a more sophisticated analysis for tracking values.
+        // Hence, we abort in this case.
+
+        return;
+      }
+    }
+
+    if (numReturnedExpressions != 2) {
+      addIssueOnReturnedExpressions(ctx, returnStatement, String.format(INVALID_GETNEWARGSEX_ELEMENT_COUNT_MESSAGE, numReturnedExpressions));
+      return;
+    }
+
+    // Exactly two expressions are being returned
+    final Expression firstElement = returnedExpressions.get(0);
+    final Expression secondElement = returnedExpressions.get(1);
+
+    if (!firstElement.type().canBeOrExtend(BuiltinTypes.TUPLE) ||
+      !secondElement.type().canBeOrExtend(BuiltinTypes.DICT)) {
+      ctx.addIssue(firstElement.firstToken(), secondElement.lastToken(), INVALID_GETNEWARGSEX_TUPLE_MESSAGE);
+    }
+  }
+
+  private static InferredType getReturnStmtType(ReturnStatement returnStatement) {
+    final List<Expression> returnedExpressions = returnStatement.expressions();
+
+    if (returnedExpressions.isEmpty()) {
+      return InferredTypes.NONE;
+    }
+
+    if (returnedExpressions.size() == 1) {
+      return returnedExpressions.get(0).type();
+    }
+
+    return InferredTypes.TUPLE;
+  }
+
+  /**
+   * Calls {@code ctx.addIssue} for a return statement such that...
+   *
+   * ...all returned expressions are marked as the source of the issue if the return statement contains such expressions
+   * ...the return keyword is marked as the source of the issue if the return statement does not contain any expressions
+   */
+  private static void addIssueOnReturnedExpressions(SubscriptionContext ctx, ReturnStatement returnStatement, String message) {
+    final List<Expression> returnedExpressions = returnStatement.expressions();
+
+    if (returnedExpressions.isEmpty()) {
+      ctx.addIssue(returnStatement.returnKeyword(), message);
+    } else {
```

Comment:
Same here, I think this condition is not needed because you early return al line 190 if `returnedExpressions.isEmpty()`

Reply:
In this case, I would like to keep the check for an empty expression list, if possible.
The method is used a lot in the code, and it should be safe to call if it is ever applied to a non-empty list of expressions in the future.